spinoso-string/src/enc/ascii/mod.rs

```diff
@@ -280,6 +281,18 @@ impl AsciiString {
     pub fn extend_from_slice(&mut self, other: &[u8]) {
         self.inner.extend_from_slice(other);
     }
+
+    #[inline]
+    pub fn codepoints(&self) -> Codepoints {
+        let iter = self
+            .inner
+            .as_slice()
+            .bytes()
+            .map(char::from)
+            .collect::<Vec<_>>()
+            .into_iter();
+        Codepoints::from(iter)
+    }
```

Comment:
usually I'd implement this by having each encoding have their own ` CodePoints` iterator.

Then the codepoints iterator for string would be defined like this:

```rust
pub struct Codepoints<'a> {
    state: State<'a>,
}

enum State<'a> {
    Ascii(ascii::Codepoints<'a>),
    Binary(binary::Codepoints<'a>),
    Utf8(utf8::Codepoints<'a>),
}
```

Take a look at how `Inspect` is implemented:

https://github.com/artichoke/artichoke/blob/0ae1fb15dee52dc736d05302804625a4d4c8819c/spinoso-string/src/enc/inspect.rs#L5-L35

This avoids the allocation to normalize everything to `IntoIter`.

Reply:
Thanks! This was exactly what I tried to figure out but couldn't. I reimplement with this design in b8729b7.

Reply:
awesome! new revision looks great. I kinda love this pattern of handling the various encodings and Rust's lifetime support makes the code super easy to reason about (and elegant IMO).